### ISSUE TITLE
Regex Match, Split and Matches should support RegexOptions.AnyNewLine as (?=\r\z|\n\z|\r\n\z|\z)

### ISSUE THREAD

**Issue by jzabroski** [dotnet/runtime#25598]
From [MSDN](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#Multiline):

> By default, $ matches only the end of the input string. If you specify the RegexOptions.Multiline option, it matches either the newline character (\n) or the end of the input string. It does not, however, match the carriage return/line feed character combination. To successfully match them, use the subexpression \r?$ instead of just $.

This is by far one of the biggest gotchas with using .NET Regex class.

I suggest adding a RegexOptions.AnyNewLine which treats $ as matching _both_ Windows' Environment.NewLine and UNIX' Environment.NewLine, regardless of the Environment running corefx.

*Portability concerns*
According to Wikipedia, there are a ton of different operating systems, all with different line ending settings. The current implementation hardcodes Unix line-ending style. RegexOptions.AnyNewLine, defined as (?=[\r\n]|\z), would add support for Windows line-ending style.

The advise written in the current docs is actually not portable on Unix, which is becoming a more popular option. As it is suggested, \r?$ will capture one or two lines on Unix, and one on Windows. If you try running Windows assemblies with this hack on Linux, you will change the semantics of programs.

*Backward compatibility concerns*
Fully backward compatible: This RegexOptions enum extension would not be a default, and so it would not break any clients with reasonably written code. The only existing code that might display different behavior would be reflection code that sets every option on RegexOptions enum variable. I really can't envision anyone doing this on purpose.

Here is Petr Onderka (@svick)'s summary: 

OS | Line-ending style | Current | Environment.NewLine | AnyNewLine
-- | -- | -- | -- | --
Windows | Windows | ✗ | ✓ | ✓
Windows | Unix | ✓ | ✗ | ✓
Unix | Windows | ✗ | ✗ | ✓
Unix | Unix | ✓ | ✓ | ✓

## Api Proposal
edit by @ViktorHofer

```diff
namespace System.Text.RegularExpressions
{
    [Flags]
    public enum RegexOptions
    {
        None                    = 0x0000,
        IgnoreCase              = 0x0001, // "i"
        Multiline               = 0x0002, // "m"
        ExplicitCapture         = 0x0004, // "n"
        Compiled                = 0x0008, // "c"
        Singleline              = 0x0010, // "s"
        IgnorePatternWhitespace = 0x0020, // "x"
        RightToLeft             = 0x0040, // "r"

#if DEBUG
        Debug                   = 0x0080, // "d"
#endif

        ECMAScript              = 0x0100, // "e"
        CultureInvariant        = 0x0200,
+       AnyNewLine              = 0x0400 // Treat "$" as (?=[\r\n]|\z)
    }
}
```

## [API Review Notes](https://github.com/dotnet/corefx/issues/28410#issuecomment-392863409)

[Video](https://www.youtube.com/watch?v=ZHKLi8qWTCs&t=0h0m0s)

Looks good. A few comments:

* ~~We cannot use the proposed value of 128 because it's already taken (see #if DBG in code)~~ Spec updated so that `AnyNewLine = 0x400` (1024).
* ~~The table looks wrong (Windows on Windows on the Current should work IMHO)~~ The table is correct. The fact this trips up experts just speaks to why this is a profound GOTCHA in the Core SDK.
* ~~May be AcceptAllLineEndings?~~ Some hallway testing I've done indicates AnyNewLine is a good name.  Plus, (argument after the final name was chosen) this enumeration will be transliterated into a checkbox on Regular Expression Visualization tools like Regex Hero, so it is preferable to have a concise explanation for the feature to avoid excessive screen space.

## PR Review Notes
After work had started on the approved proposal, @danmosemsft [asked if the scope of this feature should be changed](https://github.com/dotnet/corefx/pull/41195#issuecomment-538600562) to also adjust the meaning `\Z` anchor.  @jzabroski suggested writing how the end user documentation will look after this change, as good docs will determine if it is a function step improvement in usability and reducing gotchas.

Also, during the PR, it seems @shishirchawla also proposed AnyEndZ as a way to use AnyNewLine as an "Anchor Modifier", which will alter the meaning of '\Z' anchor **in addition to altering the meaning of '$' anchor**.  The intent of this improvement appears to be to remove all platform-specific language from the [Anchors documentation](https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions), which seems like a great improvement.

# AnyNewLine as Anchor Modifier to \Z and $ Anchors

| flags | $ is treated as | $ documentation | \Z is treated as | \Z documentation |
| ----- | --------------- | ------------------- | ----------------- | -------------------- |
| neither | `(?=\n\z\|\z)` | The match must occur at the end of the string or before `\n` at the end of the string. | (Same as `$` with this option.) | (Same as `$` with this option.) |
| `RegexOptions.Multiline` | `(?=\n\|\n\z\|\z)` | The match must occur at the end of the string or before `\n` anywhere in the string. | `(?=\n\z\|\z)` | The match must occur at the end of the string or before `\n` at the end of the string. |
| `RegexOptions.Multiline \| RegexOptions.AnyNewLine` | `(?=\r\n\|\r\|\n\|\r\n\z\|\r\z\|\n\z\|\z)` | The match must occur at the end of the string or before `\r\n`, `\n` or `\r` anywhere in the string. | `(?=\r\n\z\|\r\z\|\n\z\|\z)` | The match must occur at the end of the string or before `\r\n`, `\n` or `\r` at the end of the string. |
| `RegexOptions.AnyNewLine` | `(?=\r\n\z\|\r\z\|\n\z\|\z)` | The match must occur at the end of the string or before `\r\n`, `\n` or `\r` at the end of the string. | (Same as `$` with this option.) | (Same as `$` with this option.) |

**Comment by danmosemsft** [comment]
I was working to rebase #1449 on master, but it needs partly reimplementing at this point. And I realized there are some omissions, but a larger issue is that adding support in the same way for the other 3-5 kinds of newlines in [TR18#1.6](http://www.unicode.org/reports/tr18/#RL1.6) will make the code unwieldy. Currently `\n` is already specially treated in 13 places in the interpreter, compiler, and each of those would need an update to recognize the other line endings, lengthening hot code paths, with on top of that the special complication that `\r\n` is two characters and should be treated as one.

I thought about this for some time and realized that we could instead "lower" several of these into equivalent primitive components, so that nothing but the parser need know about them. Here is how they "lower", as best I can determine:


| flags                                              | ^ lowered to (no capture)                           | $ lowered to (no capture)                              | \Z lowered to (no capture)                   |
|----------------------------------------------------|-----------------------------------------------------|--------------------------------------------------------|----------------------------------------------|
| neither                                            | `\A`                                                | `(?=\z\|\n\z)`                                          | (Same as `$` with neither)                   |
| `RegexOptions.AnyNewLine` (new)                         | `\A`                                                | `(?=\r\n\z\|[\r\u0085\u2028\u2029]?\z)\|(?<!\r)(?=\n\z)` | (Same as `$` with `RegexOptions.AnyNewLine`) |
| `RegexOptions.Multiline`                           | `(?<=\A\|\n)`                                        | `(?=\z\|\n)`                                            | (Same as `$` with neither)                   |
| `RegexOptions.Multiline \| RegexOptions.AnyNewLine` (new) | `(?<=\A\|\r\n\|[\n\u0085\u2028\u2029])\|(?<=\r)(?!\n)` | `(?=\r\n\|[\r\u0085\u2028\u2029]\|\z)\|(?<!\r)(?=\n)`     | (Same as `$` with `RegexOptions.AnyNewLine`) |

| flags                                               | . lowered to (it captures)  |
|-----------------------------------------------------|-----------------------------|
| neither                                             | `[^\n\r\u0085\u2028\u2029]` |
| `RegexOptions.SingleLine`                           | `.`                         |
| `RegexOptions.SingleLine \| RegexOptions.AnyNewLine` (new) | `(\r\n\|(?!\r\n).)`          |

The above encapsulates some special behaviors:

*  `$` and `.` (assuming AnyNewLine) must match only once in `\r\n`, but otherwise everything else is fair game - eg., `\r\r` is two matches.
*  without Multiline, `$` and `^` are asymmetrical. `$` matches "abc\n" twice - both at the ` \n` and the end (same as PCRE, but not Python) whereas without Multiline `^` only matches "\nabc" one time.

TR18#1.6 also recommends a `\R` to match any newline. That's a separate proposal, but it could be easily accommodated in the parser by lowering it:

| flags | \R (if added) lowered to    (it captures)        |
|-------|--------------------------------------------------|
| any   | `(\r\n\|(?!\r\n)[\n\r\u0085\u2028\u2029])`        |

Note this is not the same as the example pattern in TR18.

All the above excludes '\v' and '\f': it's clear from TR18 they MAY match, but it's not clear whether they MUST. The neat thing about lowering is that if we want them, it's just a matter of adding them next to `\u0085` etc in the parser and not in 13 other places.

Of course, for efficiency reasons we probably want to leave today's special knowledge of line endings in the runners, because we probably can't currently evaluate `(?=\z\|\n\z)` as efficiently as a simple length and single character check. But the new AnyNewLine behavior could lower, as it's probably OK if it's not quite as fast. The performance without AnyNewLine would not be affected.

In short, I think we should start again and this time only touch the parser.

Thoughts -- @stephentoub @jzabroski ?

**Comment by svick** [comment]
@danmosemsft 

>  But the new AnyNewLine behavior could lower, as it's probably OK if it's not quite as fast. […] Later, we could look at optimizing […]

I think that the speed of `AnyNewLine` should be on par with the current behavior, at least as long as we're talking about the common case of dealing only with `\n` and `\r\n`. Would it be reasonable to somehow detect the presence of the less common line endings and only switch to the less efficient approach if they are found?

Or are you saying that the performance of `AnyNewLine` with your proposed implementation *could* be on par with the current behavior, but it's not necessary for the initial implementation to get there?

**Comment by danmosemsft** [comment]
> that the performance of AnyNewLine with your proposed implementation could be on par with the current behavior, but it's not necessary for the initial implementation to get there?

This, although it's never going to be possible to be as good given there's necessarily more tests and branches.

**Comment by jzabroski** [comment]
@danmosemsft Thanks for putting this together. This issue has been on my mind for some time, but life keeps getting in the way!

The main reason I did not put this on my plate is precisely the risk of causing a regression in behavior that you highlight by the thirteen different places that need to be considered.  Perhaps I am too detail-oriented, but as we are nearing .NET Core 5, my thought was, "Maybe this should wait for the next minor version?"

In terms of your design proposal, I agree that making the code long-term easier to understand and maintain is a big win.  Having the parser handle the phrase as a token makes sense, and also cuts down the risk of a major regression due to unforeseen feature combinations.  It would also be easier to lift back up to the interpreter / compiler the token.  Plus, you could have a compiler that analyzes the parse tree to determine if it is an "extended Regular Expression" or actual "regular language Regular Expression."  Similar to how .NET Expression tree API allows preferInterpretation.  In most cases, it would be convenient to have the high-level language services just transparently do what's optimal.

**Comment by jzabroski** [comment]
@stephentoub  For .NET vNext, can you comment on @danmosemsft ideas to lift this: https://github.com/dotnet/runtime/issues/25598#issuecomment-666955410

**Comment by stephentoub** [comment]
I like the simplicity and isolated nature of that approach, and I think it's reasonable to start, at least for a prototype, and see a) whether it's functionally correct with all the various new tests we throw at it, and b) what the performance hit is without further optimization.  Assuming it's functionally sound and the perf hit is reasonable, it sounds like a good way to go for now; if in the future the perf of it became more of an issue, we could address that in a variety of ways, either by adding more optimizations for the patterns it employs, or by switching away from doing it purely in the parser as a lowering step.

**Comment by danmosemsft** [comment]
Now we just need a volunteer...

**Comment by jzabroski** [comment]
(Just got back from vacation.) I'm interested in doing it, and think I almost got my two open source projects in a state where they're ready for v5.  Anyway, I think it would be a lot of fun to do - if I use that word. :)

**Comment by danmosemsft** [comment]
Great, I assigned you @jzabroski . As @stephentoub mentions, you'll probably want to start by prototyping and getting perf measurements. Thanks!

**Comment by danmoseley** [comment]
@jzabroski I cleared your assignment.

**Comment by jzabroski** [comment]
@danmoseley Thanks. Sorry about not finishing this. I caught Covid shortly after you assigned this and my time for additional work outside my normal hours diminished (long recovery).

**Comment by danmoseley** [comment]
no problem! sorry to hear about that 😿 you're always welcome to contribute here or to anything else in the repo.

**Comment by danmoseley** [comment]
Returning to this briefly, I wanted to point out that the behavior described in the top post is not what we want. It says that "The match must occur at the end of the string or before \r\n, \n or \r at the end of the string."

That is not desirable, because `\r\n` is logically a single line break. This how it is treated by all text editors, that is what is intended when code (usually on Windows) writes it. Also, the Unicode regex standard [TR18](https://www.unicode.org/reports/tr18/#RL1.6) explicitly indicates that they go together indivisibly. And it is the behavior of PCRE, Perl, Java etc when you use \R.

That's why in my table above, I have $ equivalent to (effectively) `(?=\r\n\z|\r?\z)|(?<!\r)(?=\n\z)` (in non multiline) and not `(?=\r\n\z|\r\z|\n\z|\z)` as in the topmost post.

**Comment by danmoseley** [comment]
@stephentoub I hacked up what lowering might look like -- see https://github.com/dotnet/runtime/compare/main...danmoseley:runtime:anynewline?expand=1#diff-548465bb4dbdd05024f0676c944e3c2b89677c48257bf28dfb0afccde31ded0aR19

I wrapped the pattern in an iterator with the idea that it could show "real" locations in the pattern in error messages, but not sure whether it's worth that abstraction.

Do you think this is a worthwhile direction?

**Comment by danmoseley** [comment]
I didn't like the parser approach -- it's clumsy and will run slow.
I'm working on supporting this properly, but in the generator only -- that way we can do it efficiently plus avoid adding any runtime cost for the regular case.

working in  https://github.com/dotnet/runtime/compare/main...danmoseley:anynewline2?expand=1

**Comment by stephentoub** [comment]
> I'm working on supporting this properly, but in the generator only

You mean someone else would add it for the other engines?

**Comment by danmoseley** [comment]
Potentially, in future, with care to avoid impacting perf in existing cases. As AnyNewLine would be new, I think it's OK if only the new engine supports it (meaning it will only work for static patterns) initially.

Just checking for more than \n is easy, the complexity is dealing with \r\n atomically

**Comment by danmoseley** [comment]
What are your thoughts, would it be acceptable to only support for static patterns (generator)

**Comment by stephentoub** [comment]
> I think it's OK if only the new engine supports it (meaning it will only work for static patterns)

I'm not a fan of that. At a minimum, when we support it, it should be in all but the NonBacktracking engine, but ideally all of them.

**Comment by gregli-msft** [comment]
Reviving an old thread, I ran across this work in adding regular expressions to [Power Fx](https://github.com/microsoft/Power-Fx/blob/main/docs/overview.md). We run on top of .NET and JavaScript and potentially other platforms in the future, so I'm trying to carve out a regular expression language that can either be lowered/transpiled and/or limited in expressiveness for consistency. Consistency with Excel's use of PCRE2 is important too. We've been successfully using [XRegExp](https://xregexp.com/) for many years in Power Apps so we are comfortable with the lowering approach. We can also be pretty opinionated about what we support, for example we are "no capture" by default and named and numbered captures can't be used together, partially because .NET and JavaScript treat mixing them differently, thus we effectively guide toward the better named captures.

One area of inconsistency that we need to resolve is newline behavior, the subject of this issue. Power Apps already uses `\r\n` for newlines so I need that combination to work. I'm in the process of integrating [the lowering formulas](https://github.com/microsoft/Power-Fx/blob/587eff33a4afe13dcea5d5a9d36cfc80c7915382/src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryRegEx.cs#L243) that @danmoseley posted earlier (thank you!) and they are working well.  

I have a few points/questions:

1. **Dot divisible**: Dan brought up that `\r\n` should be indivisible and provided a lowering formula for `.` to make that work. I know that [TR18#1.6](http://www.unicode.org/reports/tr18/#RL1.6) states "(CRLF) _should_ match as if it were a single character" (their emphasis on "should") and they go on to say that it isn't required for conformance. In my research, I have been unable to find any regular expression engine that supports `\r\n` as a newline combination that does this. For me, it is odd that `.` could match anything but one character, that seems like a fundamental assumption. 
 
    Practically, if I understand the requirement, `/%.%/s` should match `%\r\n%`. _Am I misunderstanding?_  
 
    For example, both fail as written here, but succeed with `\r` or `\n` alone.
    - JavaScript: `console.log( "%\r\n%".match( /%.%/s ) )` with [JSFiddle](https://jsfiddle.net/bfv3L5gu/).
    - Excel, using PCRE2: `=REGEXTEST("%" & CHAR(13) & CHAR(10) & "%", "(?s)%.%" )`.  

1. **Caret and Dollar indivisible**: I do think it makes sense for `\r\n` to be treated as one for these anchors. The test case that swayed me is `/(%|)$/mg` with `%\r\n%`. On JavaScript, that will result in 3 matches: two for "a" and a third for the empty match after the `\r` but before the `\n` (see the JSFiddle). 

    PCRE2 only returns the two `a` results which seems correct, in Excel `=REGEXEXTRACT( "%" & CHAR(13) & CHAR(10) & "%", "(?m)^(%|)", 1 )`. 

    No question here, I assume this is still the direction you all are headed, and what I plan to do with Power Fx.  I just need to figure out how to do it in JavaScript without `\A` or `\z` support.

1. **Newline characters**: This thread started out specifically about adding `\r`. [TR18#1.6](http://www.unicode.org/reports/tr18/#RL1.6) calls out 7 newline characters, not including the `\r\n` combination. Dan's lowering supported 5 characters, dropping `\v` and `\f` (somewhat dated?). For reference, PCRE2 has two relevant modes, `ANYCRLF` with any combination of `\r` and `\n` which is what Excel is using, and `ANY` which supports all 7 Unicode newline characters.  JavaScript supports 4 , the 5 Dan selected minus `\x85`.

   It may be impossible to tell what .NET will do until it is done. Still, we have to make a choice for Power Fx soon. I can see the simplicity of just adding `\r` as that is what is really needed, I can see Dan's compromise, and I can see respecting the full Unicode set. If we pursue lowering for .NET, which would be engaged only when this new option is selected, I don't know that additional characters are that much more expensive. If it is hand coded in a dozen or so places, and it impacts all users with or without the option selected, I can see going small.  

    _Any thoughts or guidance on which way this may go?_

Thanks for all the help!  I'd love to see this work continue, but like all of you I am short on spare time. I'm happy to help however I can when this initiative does get rolling again.
